vllm/v1/core/sched/scheduler.py

```diff
@@ -650,6 +646,17 @@ def update_from_output(
                 request.structured_output_request.grammar.accept_tokens(  # type: ignore[union-attr]
                     req_id, new_token_ids)
 
+            # Add newly generated spec token ids to the request.
+            if spec_token_ids is not None:
+                if request.use_structured_output:
+                    metadata = request.structured_output_request
+                    assert metadata is not None and metadata.grammar is not None
+                    # Needs to happen after new_token_ids are accepted.
+                    request.spec_token_ids = metadata.grammar.validate_tokens(
+                        spec_token_ids[req_index])
+                else:
+                    request.spec_token_ids = spec_token_ids[req_index]
+
```

Comment:
Should this logic come before the changes made above ?
```
grammar_bitmask = self.structured_output_manager.grammar_bitmask(
            self.requests,
            structured_output_request_ids,
            len(self.running),
            scheduled_spec_decode_tokens,
        )
```

My understanding is that currently, this PR is creating bitmask based on the all K draft tokens and then rejects R tokens among the K tokens which do not follow the schema so we end up doing fwd pass on K-R tokens only. If we bring this logic above then we need to create only (K-R) bitmasks?

Reply:
Same response as above, since this is happening in `update_from_output` and not `schedule()`, the bitmask has not yet been created for these proposed tokens.